xz image decompress support

### Terraform Version
Terraform v1.4.6

### Affected Resource(s)
- images_image_v2

### Expected Behavior
Download and decompress cloud images compressed with xz, e.g. Fedora CoreOS (see OpenStack image on Fedora [download page](https://fedoraproject.org/it/coreos/download/?stream=stable)).

### Actual Behavior
Cannot decompress xz images.

### Steps to Reproduce
Just try to add `decompress = true`.

Yeah by setting `TF_LOG=TRACE` I can see that it cannot download the image, but I cannot figure out why. I tried downloading from the browser and I have no issues.

Log:
```
2023-06-14T10:56:34.707Z [INFO]  provider.terraform-provider-openstack: 2023/06/14 10:56:34 [DEBUG] File doens't exists .image_cache/407f83df289f796b1510d4d74965bd0c.img. will download from https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-GenericCloud-9-latest.x86_64.qcow2: timestamp=2023-06-14T10:56:34.707Z
2023-06-14T10:56:34.707Z [INFO]  provider.terraform-provider-openstack: 2023/06/14 10:56:34 [DEBUG] OpenStack Request URL: GET https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-GenericCloud-9-latest.x86_64.qcow2: timestamp=2023-06-14T10:56:34.707Z
2023-06-14T10:56:34.707Z [INFO]  provider.terraform-provider-openstack: 2023/06/14 10:56:34 [DEBUG] OpenStack Request Headers:
Cache-Control: no-cache: timestamp=2023-06-14T10:56:34.707Z
2023-06-14T10:56:35.188Z [INFO]  provider.terraform-provider-openstack: 2023/06/14 10:56:35 [DEBUG] OpenStack connection error, retries exhausted. Aborting: timestamp=2023-06-14T10:56:35.187Z
2023-06-14T10:56:35.188Z [TRACE] provider.terraform-provider-openstack: Called downstream: tf_provider_addr=registry.terraform.io/terraform-provider-openstack/openstack tf_req_id=8c850196-6237-3532-028e-4dc471c6be3a tf_resource_type=openstack_images_image_v2 tf_rpc=ApplyResourceChange @caller=/home/vscode/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:838 @module=sdk.helper_schema timestamp=2023-06-14T10:56:35.187Z
2023-06-14T10:56:35.189Z [TRACE] provider.terraform-provider-openstack: Received downstream response: diagnostic_error_count=1 tf_req_duration_ms=2567 tf_rpc=ApplyResourceChange @caller=/home/vscode/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tf5serverlogging/downstream_request.go:37 diagnostic_warning_count=0 tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/terraform-provider-openstack/openstack tf_req_id=8c850196-6237-3532-028e-4dc471c6be3a tf_resource_type=openstack_images_image_v2 @module=sdk.proto timestamp=2023-06-14T10:56:35.188Z
2023-06-14T10:56:35.189Z [ERROR] provider.terraform-provider-openstack: Response contains error diagnostic: diagnostic_detail= diagnostic_severity=ERROR diagnostic_summary="Error opening file for Image: Error downloading image from "https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-GenericCloud-9-latest.x86_64.qcow2"" tf_proto_version=5.3 @caller=/home/vscode/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto tf_provider_addr=registry.terraform.io/terraform-provider-openstack/openstack tf_req_id=8c850196-6237-3532-028e-4dc471c6be3a tf_resource_type=openstack_images_image_v2 tf_rpc=ApplyResourceChange timestamp=2023-06-14T10:56:35.188Z
2023-06-14T10:56:35.189Z [TRACE] provider.terraform-provider-openstack: Served request: tf_req_id=8c850196-6237-3532-028e-4dc471c6be3a tf_resource_type=openstack_images_image_v2 @caller=/home/vscode/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:831 @module=sdk.proto tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/terraform-provider-openstack/openstack tf_rpc=ApplyResourceChange timestamp=2023-06-14T10:56:35.188Z
2023-06-14T10:56:35.190Z [TRACE] maybeTainted: openstack_images_image_v2.fedora_coreos_image encountered an error during creation, so it is now marked as tainted
2023-06-14T10:56:35.190Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for openstack_images_image_v2.fedora_coreos_image
2023-06-14T10:56:35.190Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for openstack_images_image_v2.fedora_coreos_image
2023-06-14T10:56:35.191Z [TRACE] evalApplyProvisioners: openstack_images_image_v2.fedora_coreos_image is tainted, so skipping provisioning
2023-06-14T10:56:35.191Z [TRACE] maybeTainted: openstack_images_image_v2.fedora_coreos_image was already tainted, so nothing to do
2023-06-14T10:56:35.191Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for openstack_images_image_v2.fedora_coreos_image
2023-06-14T10:56:35.191Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for openstack_images_image_v2.fedora_coreos_image
2023-06-14T10:56:35.192Z [TRACE] statemgr.Filesystem: creating backup snapshot at terraform.tfstate.backup
2023-06-14T10:56:35.193Z [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 52
2023-06-14T10:56:35.193Z [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
2023-06-14T10:56:35.204Z [ERROR] vertex "openstack_images_image_v2.fedora_coreos_image" error: Error opening file for Image: Error downloading image from "https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-GenericCloud-9-latest.x86_64.qcow2"
2023-06-14T10:56:35.204Z [TRACE] vertex "openstack_images_image_v2.fedora_coreos_image": visit complete, with errors
2023-06-14T10:56:35.204Z [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/terraform-provider-openstack/openstack\"] (close)" errored, so skipping
2023-06-14T10:56:35.204Z [TRACE] dag/walk: upstream of "root" errored, so skipping
2023-06-14T10:56:35.204Z [TRACE] statemgr.Filesystem: have already backed up original terraform.tfstate to terraform.tfstate.backup on a previous write
2023-06-14T10:56:35.205Z [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 53
2023-06-14T10:56:35.205Z [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
╷
│ Error: Error opening file for Image: Error downloading image from "https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-GenericCloud-9-latest.x86_64.qcow2"
│ 
│   with openstack_images_image_v2.fedora_coreos_image,
│   on main.tf line 28, in resource "openstack_images_image_v2" "fedora_coreos_image":
│   28: resource "openstack_images_image_v2" "fedora_coreos_image" {
│ 
╵
```

--- a/go.mod
+++ b/go.mod
@@ -8,6 +8,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.25.0
 	github.com/mitchellh/go-homedir v1.1.0
 	github.com/stretchr/testify v1.7.2
+	github.com/ulikunitz/xz v0.5.11
 	gopkg.in/yaml.v2 v2.4.0
 )
 

--- a/openstack/images_image_v2.go
+++ b/openstack/images_image_v2.go
@@ -89,84 +89,105 @@ func resourceImagesImageV2FileProps(filename string) (int64, string, error) {
 func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.ResourceData) (string, error) {
 	if filename := d.Get("local_file_path").(string); filename != "" {
 		return filename, nil
-	} else if furl := d.Get("image_source_url").(string); furl != "" {
-		dir := d.Get("image_cache_path").(string)
-		if err := os.MkdirAll(dir, 0700); err != nil {
-			return "", fmt.Errorf("unable to create dir %s: %s", dir, err)
-		}
-		filename := filepath.Join(dir, fmt.Sprintf("%x.img", md5.Sum([]byte(furl))))
+	}
 
-		if _, err := os.Stat(filename); err != nil {
-			if !os.IsNotExist(err) {
-				return "", fmt.Errorf("Error while trying to access file %q: %s", filename, err)
-			}
-			log.Printf("[DEBUG] File doens't exists %s. will download from %s", filename, furl)
-			file, err := os.Create(filename)
-			if err != nil {
-				return "", fmt.Errorf("Error creating file %q: %s", filename, err)
-			}
-			defer file.Close()
-			client := &client.ProviderClient.HTTPClient
-			request, err := http.NewRequest("GET", furl, nil)
-			if err != nil {
-				return "", fmt.Errorf("Error create a new request")
-			}
-
-			username := d.Get("image_source_username").(string)
-			password := d.Get("image_source_password").(string)
-			if username != "" && password != "" {
-				request.SetBasicAuth(username, password)
-			}
-
-			resp, err := client.Do(request)
-			if err != nil {
-				return "", fmt.Errorf("Error downloading image from %q: %s", furl, err)
-			}
-
-			// check for credential error among other errors
-			if resp.StatusCode != http.StatusOK {
-				return "", fmt.Errorf("Error downloading image from %q, statusCode is %d", furl, resp.StatusCode)
-			}
-
-			defer resp.Body.Close()
-
-			reader := resp.Body
-			decompress := d.Get("decompress").(bool)
-
-			if decompress {
-				// If we're here "Content-Encoding" in not filled, we'll read
-				// "Content-Type" to select format
-				switch resp.Header.Get("Content-Type") {
-				case "gzip", "application/gzip":
-					reader, err = gzip.NewReader(resp.Body)
-					if err != nil {
-						return "", fmt.Errorf("Error decompressing gzip image: %s", err)
-					}
-				case "bzip2", "application/bzip2", "application/x-bzip2":
-					bz2Reader := bzip2.NewReader(resp.Body)
-					reader = io.NopCloser(bz2Reader)
-				case "xz", "application/xz", "application/x-xz":
-					xzReader, err := xz.NewReader(resp.Body)
-					if err != nil {
-						return "", fmt.Errorf("Error decompressing xz image: %s", err)
-					}
-					reader = io.NopCloser(xzReader)
-				default:
-					return "", fmt.Errorf("Error decompressing image, format %s is not supported", resp.Header.Get("Content-Type"))
-				}
-				defer reader.Close()
-			}
-
-			if _, err = io.Copy(file, reader); err != nil {
-				return "", fmt.Errorf("Error downloading image %q to file %q: %s", furl, filename, err)
-			}
-			return filename, nil
-		}
-		log.Printf("[DEBUG] File exists %s", filename)
-		return filename, nil
-	} else {
+	furl := d.Get("image_source_url").(string)
+	if furl == "" {
 		return "", fmt.Errorf("Error in config. no file specified")
 	}
+
+	dir := d.Get("image_cache_path").(string)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return "", fmt.Errorf("unable to create dir %s: %s", dir, err)
+	}
+
+	filename := filepath.Join(dir, fmt.Sprintf("%x.img", md5.Sum([]byte(furl))))
+	_, err := os.Stat(filename)
+	if err == nil {
+		log.Printf("[DEBUG] File exists %s", filename)
+		return filename, nil
+	}
+
+	if !os.IsNotExist(err) {
+		return "", fmt.Errorf("Error while trying to access file %q: %s", filename, err)
+	}
+
+	log.Printf("[DEBUG] File doens't exists %s. will download from %s", filename, furl)
+	file, err := os.Create(filename)
+	if err != nil {
+		return "", fmt.Errorf("Error creating file %q: %s", filename, err)
+	}
+	defer file.Close()
+
+	// a cleanup func to delete a failed file
+	delFile := func() {
+		if err := os.Remove(filename); err != nil {
+			log.Printf("[DEBUG] failed to cleanup the %q file: %s", filename, err)
+		}
+	}
+
+	httpClient := &client.ProviderClient.HTTPClient
+	request, err := http.NewRequest("GET", furl, nil)
+	if err != nil {
+		delFile()
+		return "", fmt.Errorf("Error creating a new request: %s", err)
+	}
+
+	username := d.Get("image_source_username").(string)
+	password := d.Get("image_source_password").(string)
+	if username != "" && password != "" {
+		request.SetBasicAuth(username, password)
+	}
+
+	resp, err := httpClient.Do(request)
+	if err != nil {
+		delFile()
+		return "", fmt.Errorf("Error downloading image from %q: %s", furl, err)
+	}
+
+	// check for credential error among other errors
+	if resp.StatusCode != http.StatusOK {
+		delFile()
+		return "", fmt.Errorf("Error downloading image from %q, statusCode is %d", furl, resp.StatusCode)
+	}
+
+	defer resp.Body.Close()
+	reader := resp.Body
+
+	decompress := d.Get("decompress").(bool)
+	if decompress {
+		// If we're here "Content-Encoding" in not filled, we'll read
+		// "Content-Type" to select format
+		switch resp.Header.Get("Content-Type") {
+		case "gzip", "application/gzip":
+			reader, err = gzip.NewReader(resp.Body)
+			if err != nil {
+				delFile()
+				return "", fmt.Errorf("Error decompressing gzip image: %s", err)
+			}
+		case "bzip2", "application/bzip2", "application/x-bzip2":
+			bz2Reader := bzip2.NewReader(resp.Body)
+			reader = io.NopCloser(bz2Reader)
+		case "xz", "application/xz", "application/x-xz":
+			xzReader, err := xz.NewReader(resp.Body)
+			if err != nil {
+				delFile()
+				return "", fmt.Errorf("Error decompressing xz image: %s", err)
+			}
+			reader = io.NopCloser(xzReader)
+		default:
+			delFile()
+			return "", fmt.Errorf("Error decompressing image, format %s is not supported", resp.Header.Get("Content-Type"))
+		}
+		defer reader.Close()
+	}
+
+	if _, err = io.Copy(file, reader); err != nil {
+		delFile()
+		return "", fmt.Errorf("Error downloading image %q to file %q: %s", furl, filename, err)
+	}
+
+	return filename, nil
 }
 
 func resourceImagesImageV2RefreshFunc(client *gophercloud.ServiceClient, id string) resource.StateRefreshFunc {

--- a/openstack/images_image_v2.go
+++ b/openstack/images_image_v2.go
@@ -17,11 +17,12 @@ import (
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/ulikunitz/xz"
 
 	"github.com/gophercloud/gophercloud"
 	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
 	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/members"
-	"github.com/ulikunitz/xz"
+	"github.com/gophercloud/utils/terraform/mutexkv"
 )
 
 func resourceImagesImageV2MemberStatusFromString(v string) images.ImageMemberStatus {
@@ -86,7 +87,7 @@ func resourceImagesImageV2FileProps(filename string) (int64, string, error) {
 	return filesize, filechecksum, nil
 }
 
-func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.ResourceData) (string, error) {
+func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.ResourceData, mutexKV *mutexkv.MutexKV) (string, error) {
 	if filename := d.Get("local_file_path").(string); filename != "" {
 		return filename, nil
 	}
@@ -101,7 +102,12 @@ func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.Reso
 		return "", fmt.Errorf("unable to create dir %s: %s", dir, err)
 	}
 
-	filename := filepath.Join(dir, fmt.Sprintf("%x.img", md5.Sum([]byte(furl))))
+	// calculate the hashsum and create a lock to prevent simultaneous file access
+	md5sum := fmt.Sprintf("%x", md5.Sum([]byte(furl)))
+	mutexKV.Lock(md5sum)
+	defer mutexKV.Unlock(md5sum)
+
+	filename := filepath.Join(dir, fmt.Sprintf("%s.img", md5sum))
 	// a cleanup func to delete a failed file
 	delFile := func() {
 		if err := os.Remove(filename); err != nil {
@@ -115,6 +121,7 @@ func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.Reso
 	}
 
 	// check if the file size is zero
+	// it could be a leftover from older provider versions
 	if info != nil {
 		if info.Size() != 0 {
 			log.Printf("[DEBUG] File exists %s", filename)

--- a/openstack/images_image_v2.go
+++ b/openstack/images_image_v2.go
@@ -21,6 +21,7 @@ import (
 	"github.com/gophercloud/gophercloud"
 	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
 	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/members"
+	"github.com/ulikunitz/xz"
 )
 
 func resourceImagesImageV2MemberStatusFromString(v string) images.ImageMemberStatus {
@@ -119,7 +120,7 @@ func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.Reso
 
 			resp, err := client.Do(request)
 			if err != nil {
-				return "", fmt.Errorf("Error downloading image from %q", furl)
+				return "", fmt.Errorf("Error downloading image from %q: %s", furl, err)
 			}
 
 			// check for credential error among other errors
@@ -144,6 +145,12 @@ func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.Reso
 				case "bzip2", "application/bzip2", "application/x-bzip2":
 					bz2Reader := bzip2.NewReader(resp.Body)
 					reader = io.NopCloser(bz2Reader)
+				case "xz", "application/xz", "application/x-xz":
+					xzReader, err := xz.NewReader(resp.Body)
+					if err != nil {
+						return "", fmt.Errorf("Error decompressing xz image: %s", err)
+					}
+					reader = io.NopCloser(xzReader)
 				default:
 					return "", fmt.Errorf("Error decompressing image, format %s is not supported", resp.Header.Get("Content-Type"))
 				}

--- a/openstack/images_image_v2.go
+++ b/openstack/images_image_v2.go
@@ -102,14 +102,26 @@ func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.Reso
 	}
 
 	filename := filepath.Join(dir, fmt.Sprintf("%x.img", md5.Sum([]byte(furl))))
-	_, err := os.Stat(filename)
-	if err == nil {
-		log.Printf("[DEBUG] File exists %s", filename)
-		return filename, nil
+	// a cleanup func to delete a failed file
+	delFile := func() {
+		if err := os.Remove(filename); err != nil {
+			log.Printf("[DEBUG] failed to cleanup the %q file: %s", filename, err)
+		}
 	}
 
-	if !os.IsNotExist(err) {
+	info, err := os.Stat(filename)
+	if err != nil && !os.IsNotExist(err) {
 		return "", fmt.Errorf("Error while trying to access file %q: %s", filename, err)
+	}
+
+	// check if the file size is zero
+	if info != nil {
+		if info.Size() != 0 {
+			log.Printf("[DEBUG] File exists %s", filename)
+			return filename, nil
+		}
+		// delete the zero size file
+		delFile()
 	}
 
 	log.Printf("[DEBUG] File doens't exists %s. will download from %s", filename, furl)
@@ -118,13 +130,6 @@ func resourceImagesImageV2File(client *gophercloud.ServiceClient, d *schema.Reso
 		return "", fmt.Errorf("Error creating file %q: %s", filename, err)
 	}
 	defer file.Close()
-
-	// a cleanup func to delete a failed file
-	delFile := func() {
-		if err := os.Remove(filename); err != nil {
-			log.Printf("[DEBUG] failed to cleanup the %q file: %s", filename, err)
-		}
-	}
 
 	httpClient := &client.ProviderClient.HTTPClient
 	request, err := http.NewRequest("GET", furl, nil)

--- a/openstack/resource_openstack_images_image_v2.go
+++ b/openstack/resource_openstack_images_image_v2.go
@@ -302,7 +302,7 @@ func resourceImagesImageV2Create(ctx context.Context, d *schema.ResourceData, me
 		var imgFile *os.File
 
 		// downloading/getting image file props
-		imgFilePath, err = resourceImagesImageV2File(imageClient, d)
+		imgFilePath, err = resourceImagesImageV2File(imageClient, d, config.MutexKV)
 		if err != nil {
 			return diag.Errorf("Error opening file for Image: %s", err)
 		}

--- a/openstack/resource_openstack_images_image_v2.go
+++ b/openstack/resource_openstack_images_image_v2.go
@@ -173,13 +173,14 @@ func resourceImagesImageV2() *schema.Resource {
 				Type:          schema.TypeBool,
 				Optional:      true,
 				ForceNew:      false,
-				ConflictsWith: []string{"local_file_path", "verify_checksum"},
+				ConflictsWith: []string{"local_file_path", "verify_checksum", "decompress"},
 			},
 
 			"decompress": {
-				Type:     schema.TypeBool,
-				Optional: true,
-				ForceNew: true,
+				Type:          schema.TypeBool,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"web_download"},
 			},
 
 			// Computed-only
@@ -279,7 +280,21 @@ func resourceImagesImageV2Create(ctx context.Context, d *schema.ResourceData, me
 
 	var fileChecksum string
 	useWebDownload := d.Get("web_download").(bool)
-	if !useWebDownload {
+	if useWebDownload {
+		// import
+		imgURL := d.Get("image_source_url").(string)
+
+		importOpts := &imageimport.CreateOpts{
+			Name: imageimport.WebDownloadMethod,
+			URI:  imgURL,
+		}
+
+		log.Printf("[DEBUG] Import Options: %#v", importOpts)
+		res := imageimport.Create(imageClient, d.Id(), importOpts)
+		if res.Err != nil {
+			return diag.Errorf("Error while importing url %q: %s", imgURL, res.Err)
+		}
+	} else {
 		// variable declaration
 		var err error
 		var imgFilePath string
@@ -307,20 +322,6 @@ func resourceImagesImageV2Create(ctx context.Context, d *schema.ResourceData, me
 		res := imagedata.Upload(imageClient, d.Id(), imgFile)
 		if res.Err != nil {
 			return diag.Errorf("Error while uploading file %q: %s", imgFilePath, res.Err)
-		}
-	} else {
-		// import
-		imgURL := d.Get("image_source_url").(string)
-
-		importOpts := &imageimport.CreateOpts{
-			Name: imageimport.WebDownloadMethod,
-			URI:  imgURL,
-		}
-
-		log.Printf("[DEBUG] Import Options: %#v", importOpts)
-		res := imageimport.Create(imageClient, d.Id(), importOpts)
-		if res.Err != nil {
-			return diag.Errorf("Error while importing url %q: %s", imgURL, res.Err)
 		}
 	}
 

